### PR TITLE
New logging

### PR DESCRIPTION
**Pull request by melekes**
Opening for review.

### CODE REVIEW

**Comment by melekes** [comment]
```
I[05-12|17:46:30.846] Loaded PrivValidator                        module=main module=node file=/home/vagrant/.tendermint/priv_validator.json privValidator="PrivValidator{F40FC4126F05C8AC41672878638BE73B25632514 LH:87, LR:0, LS:3}"
I[05-12|17:46:30.988] Starting Node                               module=main module=node impl=Node
I[05-12|17:46:30.995] Local listener                              module=main module=node module=p2p ip=:: port=46656
I[05-12|17:46:30.996] Getting UPNP external address               module=main module=node module=p2p
I[05-12|17:46:33.999] Could not perform UPNP discover             module=main module=node module=p2p error="write udp4 0.0.0.0:32925->239.255.255.250:1900: i/o timeout"
I[05-12|17:46:34.002] Starting DefaultListener                    module=main module=node module=p2p impl=Listener(@10.0.2.15:46656)
I[05-12|17:46:34.004] Starting ConsensusState                     module=main module=node module=consensus impl=ConsensusState
I[05-12|17:46:34.010] Catchup by replaying consensus messages     module=main module=node module=consensus height=88
I[05-12|17:46:34.014] Replay: New Step                            module=main module=node module=consensus height=88 round=0 step=RoundStepNewHeight
I[05-12|17:46:34.015] Replay: Done                                module=main module=node module=consensus
I[05-12|17:46:34.017] Starting RPC HTTP server on tcp socket 0.0.0.0:46657module=main module=node module=rpcserver
I[05-12|17:46:34.020] enterNewRound(88/0). Current: 88/0/RoundStepNewHeightmodule=main module=node module=consensus
I[05-12|17:46:34.020] enterPropose(88/0). Current: 88/0/RoundStepNewRoundmodule=main module=node module=consensus
I[05-12|17:46:34.029] Signed proposal                             module=main module=node module=consensus height=88 round=0 proposal="Proposal{88/0 1:B5F4F74924CD (-1,:0:000000000000) /769AA58EA148.../}"
I[05-12|17:46:34.032] Received complete proposal block            module=main module=node module=consensus height=88 hash=F00324EA2BFDFACE69357A1CF1D18A81B6514E22
I[05-12|17:46:34.032] enterPrevote(88/0). Current: 88/0/RoundStepProposemodule=main module=node module=consensus
I[05-12|17:46:34.034] Signed and pushed vote                      module=main module=node module=consensus height=88 round=0 vote="Vote{0:F40FC4126F05 88/00/1(Prevote) F00324EA2BFD /9AEA3A6933F7.../}" error=null
I[05-12|17:46:34.036] Added to prevote                            module=main module=node module=consensus vote="Vote{0:F40FC4126F05 88/00/1(Prevote) F00324EA2BFD /9AEA3A6933F7.../}" prevotes="VoteSet{H:88 R:0 T:1 +2/3:F00324EA2BFDFACE69357A1CF1D18A81B6514E22:1:B5F4F74924CD BA{1:X} map[]}"
I[05-12|17:46:34.037] enterPrecommit(88/0). Current: 88/0/RoundStepPrevotemodule=main module=node module=consensus
I[05-12|17:46:34.039] Signed and pushed vote                      module=main module=node module=consensus height=88 round=0 vote="Vote{0:F40FC4126F05 88/00/2(Precommit) F00324EA2BFD /93C0EE221324.../}" error=null
I[05-12|17:46:34.039] Added to precommit                          module=main module=node module=consensus vote="Vote{0:F40FC4126F05 88/00/2(Precommit) F00324EA2BFD /93C0EE221324.../}" precommits="VoteSet{H:88 R:0 T:2 +2/3:F00324EA2BFDFACE69357A1CF1D18A81B6514E22:1:B5F4F74924CD BA{1:X} map[]}"
I[05-12|17:46:34.040] enterCommit(88/0). Current: 88/0/RoundStepPrecommitmodule=main module=node module=consensus
I[05-12|17:46:34.041] Block{
  Header{
    ChainID:        test-chain-ZQHh72
    Height:         88
    Time:           2017-05-12 17:46:34.024 +0000 UTC
    NumTxs:         0
    LastBlockID:    42A429D8FB16584C79B500B2D9DD4E68AC11EBBA:1:2A34667BD93F
    LastCommit:     7B766ED546954F8266A40D5B2F5938EE446F2BD8
    Data:
    Validators:     80248C0F6B58C7D4E32A9D273A0B75772982C5A4
    App:
  }#F00324EA2BFDFACE69357A1CF1D18A81B6514E22
  Data{

  }#
  Commit{
    BlockID:    42A429D8FB16584C79B500B2D9DD4E68AC11EBBA:1:2A34667BD93F
    Precommits: Vote{0:F40FC4126F05 87/00/2(Precommit) 42A429D8FB16 /35DE4B52CF83.../}
  }#7B766ED546954F8266A40D5B2F5938EE446F2BD8
}#F00324EA2BFDFACE69357A1CF1D18A81B6514E22module=main module=node module=consensus
```

multiple key is fixed in https://github.com/tendermint/tmlibs/pull/11

File: cmd/tendermint/commands/root.go [hunk 1]
```diff
@@ -24,7 +27,18 @@ var RootCmd = &cobra.Command{
 		err := viper.Unmarshal(config)
 		config.SetRoot(config.RootDir)
 		cfg.EnsureRoot(config.RootDir)
-		logger.SetLogLevel(config.LogLevel)
+		var option log.Option
+		switch config.LogLevel {
```
Comment (ebuchman) [inline cmd/tendermint/commands/root.go]: lets move to tmlibs

File: consensus/reactor.go [hunk 2]
```diff
@@ -242,12 +242,12 @@ func (conR *ConsensusReactor) Receive(chID byte, src *p2p.Peer, msgBytes []byte)
...
 		default:
 			// don't punish (leave room for soft upgrades)
-			log.Warn(Fmt("Unknown message type %v", reflect.TypeOf(msg)))
+			conR.Logger.Error(Fmt("Unknown message type %v", reflect.TypeOf(msg)))
 		}
 
 	case VoteSetBitsChannel:
 		if conR.fastSync {
-			log.Warn("Ignoring message received during fastSync", "msg", msg)
+			conR.Logger.Info("Ignoring message received during fastSync", "msg", msg)
```
Comment (ebuchman) [inline consensus/reactor.go]: Either this should be an Error or others should be an info

File: rpc/lib/client/http_client.go [hunk 3]
```diff
@@ -26,7 +26,7 @@ func makeHTTPDialer(remoteAddr string) (string, func(string, string) (net.Conn,
 	parts := strings.SplitN(remoteAddr, "://", 2)
 	var protocol, address string
 	if len(parts) != 2 {
-		log.Warn("WARNING (tendermint/rpc/lib): Please use fully formed listening addresses, including the tcp:// or unix:// prefix")
+		// log.Warn("WARNING (tendermint/rpc/lib): Please use fully formed listening addresses, including the tcp:// or unix:// prefix")
```
Comment (ebuchman) [inline rpc/lib/client/http_client.go]: Maybe its time to actually enforce this?

File: state/execution.go [hunk 4]
```diff
@@ -251,14 +252,14 @@ func (s *State) CommitStateUpdateMempool(proxyAppConn proxy.AppConnConsensus, bl
...
+		s.logger.Error("Error in proxyAppConn.CommitSync", "error", res)
 		return res
 	}
 	if res.Log != "" {
-		log.Debug("Commit.Log: " + res.Log)
+		s.logger.Debug("Commit.Log: " + res.Log)
 	}
 
-	log.Info("Committed state", "hash", res.Data)
+	s.logger.Info("Committed state", "hash", fmt.Sprintf("%X", res.Data))
```
Comment (ebuchman) [inline state/execution.go]: shouldnt need Sprintf

File: types/priv_validator.go [hunk 5]
```diff
@@ -212,7 +213,7 @@ func (privVal *PrivValidator) signBytesHRS(height, round int, step int8, signByt
 					// NOTE: proposals are non-deterministic (include time),
 					// so we can actually lose them, but will still never sign conflicting ones
 					if bytes.Equal(privVal.LastSignBytes, signBytes) {
-						log.Notice("Using privVal.LastSignature", "sig", privVal.LastSignature)
+						// log.Notice("Using privVal.LastSignature", "sig", privVal.LastSignature)
```
Comment (ebuchman) [inline types/priv_validator.go]: Maybe worth giving the privVal the logger

File: state/state.go [hunk 6]
```diff
@@ -71,6 +74,14 @@ func loadState(db dbm.DB, key []byte) *State {
 	return s
 }
 
+func (s *State) SetLogger(l log.Logger) {
+	s.logger = l
+}
+
+func (s *State) GetLogger() log.Logger {
```
Comment (ebuchman) [inline state/state.go]: is this used anywhere?

File: p2p/peer.go [hunk 7]
```diff
@@ -279,10 +279,10 @@ func (p *Peer) Get(key string) interface{} {
 }
 
 func dial(addr *NetAddress, config *PeerConfig) (net.Conn, error) {
-	log.Info("Dialing address", "address", addr)
+	// log.Info("Dialing address", "address", addr)
```
Comment (melekes) [inline p2p/peer.go]: yes. https://github.com/tendermint/tendermint/blob/feature/new-logging/p2p/switch.go#L326

File: rpc/lib/client/http_client.go (hunk 3 above)
Comment (melekes) [inline rpc/lib/client/http_client.go]: Let's do this!

File: state/state.go (hunk 6 above)
Comment (melekes) [inline state/state.go]: used it for debugging

**Comment by melekes** [comment]
```
I[05-13|08:56:27.184] Loaded PrivValidator                         module=node file=/home/vagrant/.tendermint/priv_validator.json privValidator="PrivValidator{F40FC4126F05C8AC41672878638BE73B25632514 LH:136, LR:0, LS:3}"
I[05-13|08:56:27.208] Starting multiAppConn                        module=proxy impl=multiAppConn
I[05-13|08:56:27.210] ABCI Handshake                               module=consensus appHeight=0 appHash=
I[05-13|08:56:27.211] ABCI Replay Blocks                           module=consensus appHeight=0 storeHeight=136 stateHeight=136
I[05-13|08:56:27.212] Applying block                               module=consensus height=1
I[05-13|08:56:27.215] Applying block                               module=consensus height=2
I[05-13|08:56:27.216] Applying block                               module=consensus height=3
I[05-13|08:56:27.216] Applying block                               module=consensus height=4
I[05-13|08:56:27.217] Applying block                               module=consensus height=5
I[05-13|08:56:27.218] Applying block                               module=consensus height=6
I[05-13|08:56:27.218] Applying block                               module=consensus height=7
I[05-13|08:56:27.219] Applying block                               module=consensus height=8
I[05-13|08:56:27.220] Applying block                               module=consensus height=9
I[05-13|08:56:27.220] Applying block                               module=consensus height=10
I[05-13|08:56:27.221] Applying block                               module=consensus height=11
I[05-13|08:56:27.221] Applying block                               module=consensus height=12
I[05-13|08:56:27.222] Applying block                               module=consensus height=13
I[05-13|08:56:27.223] Applying block                               module=consensus height=14
I[05-13|08:56:27.223] Applying block                               module=consensus height=15
I[05-13|08:56:27.224] Applying block                               module=consensus height=16
I[05-13|08:56:27.224] Applying block                               module=consensus height=17
I[05-13|08:56:27.225] Applying block                               module=consensus height=18
I[05-13|08:56:27.225] Applying block                               module=consensus height=19
I[05-13|08:56:27.226] Applying block                               module=consensus height=20
I[05-13|08:56:27.227] Applying block                               module=consensus height=21
I[05-13|08:56:27.227] Applying block                               module=consensus height=22
I[05-13|08:56:27.228] Applying block                               module=consensus height=23
I[05-13|08:56:27.228] Applying block                               module=consensus height=24
I[05-13|08:56:27.229] Applying block                               module=consensus height=25
I[05-13|08:56:27.229] Applying block                               module=consensus height=26
I[05-13|08:56:27.230] Applying block                               module=consensus height=27
I[05-13|08:56:27.230] Applying block                               module=consensus height=28
I[05-13|08:56:27.231] Applying block                               module=consensus height=29
I[05-13|08:56:27.232] Applying block                               module=consensus height=30
I[05-13|08:56:27.232] Applying block                               module=consensus height=31
I[05-13|08:56:27.233] Applying block                               module=consensus height=32
I[05-13|08:56:27.233] Applying block                               module=consensus height=33
I[05-13|08:56:27.234] Applying block                               module=consensus height=34
I[05-13|08:56:27.235] Applying block                               module=consensus height=35
I[05-13|08:56:27.235] Applying block                               module=consensus height=36
I[05-13|08:56:27.236] Applying block                               module=consensus height=37
I[05-13|08:56:27.237] Applying block                               module=consensus height=38
I[05-13|08:56:27.238] Applying block                               module=consensus height=39
I[05-13|08:56:27.238] Applying block                               module=consensus height=40
I[05-13|08:56:27.239] Applying block                               module=consensus height=41
I[05-13|08:56:27.240] Applying block                               module=consensus height=42
I[05-13|08:56:27.240] Applying block                               module=consensus height=43
I[05-13|08:56:27.241] Applying block                               module=consensus height=44
I[05-13|08:56:27.241] Applying block                               module=consensus height=45
I[05-13|08:56:27.242] Applying block                               module=consensus height=46
I[05-13|08:56:27.242] Applying block                               module=consensus height=47
I[05-13|08:56:27.243] Applying block                               module=consensus height=48
I[05-13|08:56:27.243] Applying block                               module=consensus height=49
I[05-13|08:56:27.244] Applying block                               module=consensus height=50
I[05-13|08:56:27.244] Applying block                               module=consensus height=51
I[05-13|08:56:27.245] Applying block                               module=consensus height=52
I[05-13|08:56:27.245] Applying block                               module=consensus height=53
I[05-13|08:56:27.246] Applying block                               module=consensus height=54
I[05-13|08:56:27.246] Applying block                               module=consensus height=55
I[05-13|08:56:27.247] Applying block                               module=consensus height=56
I[05-13|08:56:27.256] Applying block                               module=consensus height=57
I[05-13|08:56:27.257] Applying block                               module=consensus height=58
I[05-13|08:56:27.257] Applying block                               module=consensus height=59
I[05-13|08:56:27.258] Applying block                               module=consensus height=60
I[05-13|08:56:27.258] Applying block                               module=consensus height=61
I[05-13|08:56:27.258] Applying block                               module=consensus height=62
I[05-13|08:56:27.259] Applying block                               module=consensus height=63
I[05-13|08:56:27.259] Applying block                               module=consensus height=64
I[05-13|08:56:27.260] Applying block                               module=consensus height=65
I[05-13|08:56:27.260] Applying block                               module=consensus height=66
I[05-13|08:56:27.260] Applying block                               module=consensus height=67
I[05-13|08:56:27.261] Applying block                               module=consensus height=68
I[05-13|08:56:27.262] Applying block                               module=consensus height=69
I[05-13|08:56:27.262] Applying block                               module=consensus height=70
I[05-13|08:56:27.262] Applying block                               module=consensus height=71
I[05-13|08:56:27.263] Applying block                               module=consensus height=72
I[05-13|08:56:27.264] Applying block                               module=consensus height=73
I[05-13|08:56:27.264] Applying block                               module=consensus height=74
I[05-13|08:56:27.265] Applying block                               module=consensus height=75
I[05-13|08:56:27.265] Applying block                               module=consensus height=76
I[05-13|08:56:27.265] Applying block                               module=consensus height=77
I[05-13|08:56:27.265] Applying block                               module=consensus height=78
I[05-13|08:56:27.265] Applying block                               module=consensus height=79
I[05-13|08:56:27.266] Applying block                               module=consensus height=80
I[05-13|08:56:27.266] Applying block                               module=consensus height=81
I[05-13|08:56:27.266] Applying block                               module=consensus height=82
I[05-13|08:56:27.267] Applying block                               module=consensus height=83
I[05-13|08:56:27.267] Applying block                               module=consensus height=84
I[05-13|08:56:27.267] Applying block                               module=consensus height=85
I[05-13|08:56:27.267] Applying block                               module=consensus height=86
I[05-13|08:56:27.268] Applying block                               module=consensus height=87
I[05-13|08:56:27.268] Applying block                               module=consensus height=88
I[05-13|08:56:27.268] Applying block                               module=consensus height=89
I[05-13|08:56:27.268] Applying block                               module=consensus height=90
I[05-13|08:56:27.269] Applying block                               module=consensus height=91
I[05-13|08:56:27.269] Applying block                               module=consensus height=92
I[05-13|08:56:27.269] Applying block                               module=consensus height=93
I[05-13|08:56:27.269] Applying block                               module=consensus height=94
I[05-13|08:56:27.270] Applying block                               module=consensus height=95
I[05-13|08:56:27.270] Applying block                               module=consensus height=96
I[05-13|08:56:27.270] Applying block                               module=consensus height=97
I[05-13|08:56:27.270] Applying block                               module=consensus height=98
I[05-13|08:56:27.271] Applying block                               module=consensus height=99
I[05-13|08:56:27.271] Applying block                               module=consensus height=100
I[05-13|08:56:27.271] Applying block                               module=consensus height=101
I[05-13|08:56:27.271] Applying block                               module=consensus height=102
I[05-13|08:56:27.272] Applying block                               module=consensus height=103
I[05-13|08:56:27.272] Applying block                               module=consensus height=104
I[05-13|08:56:27.272] Applying block                               module=consensus height=105
I[05-13|08:56:27.272] Applying block                               module=consensus height=106
I[05-13|08:56:27.273] Applying block                               module=consensus height=107
I[05-13|08:56:27.273] Applying block                               module=consensus height=108
I[05-13|08:56:27.273] Applying block                               module=consensus height=109
I[05-13|08:56:27.273] Applying block                               module=consensus height=110
I[05-13|08:56:27.274] Applying block                               module=consensus height=111
I[05-13|08:56:27.274] Applying block                               module=consensus height=112
I[05-13|08:56:27.274] Applying block                               module=consensus height=113
I[05-13|08:56:27.274] Applying block                               module=consensus height=114
I[05-13|08:56:27.275] Applying block                               module=consensus height=115
I[05-13|08:56:27.275] Applying block                               module=consensus height=116
I[05-13|08:56:27.275] Applying block                               module=consensus height=117
I[05-13|08:56:27.275] Applying block                               module=consensus height=118
I[05-13|08:56:27.276] Applying block                               module=consensus height=119
I[05-13|08:56:27.276] Applying block                               module=consensus height=120
I[05-13|08:56:27.276] Applying block                               module=consensus height=121
I[05-13|08:56:27.276] Applying block                               module=consensus height=122
I[05-13|08:56:27.276] Applying block                               module=consensus height=123
I[05-13|08:56:27.277] Applying block                               module=consensus height=124
I[05-13|08:56:27.277] Applying block                               module=consensus height=125
I[05-13|08:56:27.277] Applying block                               module=consensus height=126
I[05-13|08:56:27.278] Applying block                               module=consensus height=127
I[05-13|08:56:27.278] Applying block                               module=consensus height=128
I[05-13|08:56:27.278] Applying block                               module=consensus height=129
I[05-13|08:56:27.278] Applying block                               module=consensus height=130
I[05-13|08:56:27.278] Applying block                               module=consensus height=131
I[05-13|08:56:27.279] Applying block                               module=consensus height=132
I[05-13|08:56:27.279] Applying block                               module=consensus height=133
I[05-13|08:56:27.279] Applying block                               module=consensus height=134
I[05-13|08:56:27.280] Applying block                               module=consensus height=135
I[05-13|08:56:27.280] Applying block                               module=consensus height=136
I[05-13|08:56:27.280] Completed ABCI Handshake - Tendermint and App are synced module=consensus appHeight=0 appHash=
I[05-13|08:56:27.286] Starting EventSwitch                         module=types impl=EventSwitch
I[05-13|08:56:27.290] Starting Node                                module=node impl=Node
I[05-13|08:56:27.291] Local listener                               module=p2p ip=:: port=46656
I[05-13|08:56:27.291] Getting UPNP external address                module=p2p
I[05-13|08:56:30.294] Could not perform UPNP discover              module=p2p error="write udp4 0.0.0.0:48509->239.255.255.250:1900: i/o timeout"
I[05-13|08:56:30.295] Starting DefaultListener                     module=p2p impl=Listener(@10.0.2.15:46656)
I[05-13|08:56:30.296] Starting P2P Switch                          module=p2p impl="P2P Switch"
I[05-13|08:56:30.297] Starting MempoolReactor                      module=consensus impl=MempoolReactor
I[05-13|08:56:30.297] Starting BlockchainReactor                   module=blockchain impl=BlockchainReactor
I[05-13|08:56:30.297] Starting ConsensusReactor                    module=consensus impl=ConsensusReactor
I[05-13|08:56:30.297] ConsensusReactor                             module=consensus fastSync=false
I[05-13|08:56:30.297] Starting ConsensusState                      module=consensus impl=ConsensusState
I[05-13|08:56:30.299] Starting WAL                                 module=consensus wal=/home/vagrant/.tendermint/data/cs.wal/wal impl=WAL
I[05-13|08:56:30.299] Starting TimeoutTicker                       module=consensus impl=TimeoutTicker
I[05-13|08:56:30.303] Catchup by replaying consensus messages      module=consensus height=137
I[05-13|08:56:30.305] Replay: New Step                             module=consensus height=137 round=0 step=RoundStepNewHeight
I[05-13|08:56:30.306] Replay: Done                                 module=consensus
I[05-13|08:56:30.307] Starting RPC HTTP server on tcp socket 0.0.0.0:46657 module=rpcserver
I[05-13|08:56:30.307] Timed out                                    module=consensus dur=-2.018236068s height=137 round=0 step=RoundStepNewHeight
I[05-13|08:56:30.308] Started node                                 module=main nodeInfo="&{PubKeyEd25519{E487062E934DC9BE6045D49105D45953A0578AB100A373643A27511F875A0150} anonymous test-chain-ZQHh72  10.0.2.15:46656 0.10.0 [wire_version=0.6.1 p2p_version=0.5.0 consensus_version=v1/0.2.2 rpc_version=0.7.0/3 tx_index=on rpc_addr=tcp://0.0.0.0:46657]}"
I[05-13|08:56:30.309] enterNewRound(137/0). Current: 137/0/RoundStepNewHeight module=consensus
```

@ebuchman should we use Debug instead of Info for "Applying block"?

**Comment by melekes** [comment]
Should we change appHash from `[]byte` to `data.Bytes`? https://github.com/tendermint/tendermint/blob/feature/new-logging/consensus/replay.go#L232

File: cmd/tendermint/commands/run_node.go [hunk 8]
```diff
@@ -84,7 +84,7 @@ func runNode(cmd *cobra.Command, args []string) error {
 	if _, err := n.Start(); err != nil {
 		return fmt.Errorf("Failed to start node: %v", err)
 	} else {
-		logger.Info("Started node", "nodeInfo", n.Switch().NodeInfo())
+		logger.Info("Started node", "nodeInfo", fmt.Sprintf("%v", n.Switch().NodeInfo()))
```
Comment (ebuchman) [inline cmd/tendermint/commands/run_node.go]: why the sprintf ?

File: cmd/tendermint/commands/root.go [hunk 9]
```diff
@@ -26,7 +27,11 @@ var RootCmd = &cobra.Command{
 		err := viper.Unmarshal(config)
 		config.SetRoot(config.RootDir)
 		cfg.EnsureRoot(config.RootDir)
-		logger = log.NewFilterByLevel(logger, config.LogLevel)
+		if config.LogLevel == "info" || config.LogLevel == "error" || config.LogLevel == "debug" {
+			logger = log.NewFilterByLevel(logger, config.LogLevel)
```
Comment (ebuchman) [inline cmd/tendermint/commands/root.go]: Would it be bad to have this logic in tmlibs too? Maybe NewFilterByLevel returns an error, or we check for nil logger?

**Comment by ebuchman** [comment]
> Should we change appHash from []byte to data.Bytes? https://github.com/tendermint/tendermint/blob/feature/new-logging/consensus/replay.go#L232

Yes. Is there ever a reason to not change things to data.Bytes ? I would think anything that might get logged or turned to json should use data.Bytes

**Comment by ebuchman** [comment]
> should we use Debug instead of Info for "Applying block"? or use something like a progress bar?

I think Info is fine, but a progress bar would be rad!

**Comment by ebuchman** [comment]
Also I think we're missing a logger in the abci client

File: cmd/tendermint/commands/run_node.go (hunk 8 above)
Comment (melekes) [inline cmd/tendermint/commands/run_node.go]: NodeInfo does not have a `String()` func ?

File: cmd/tendermint/commands/root.go (hunk 9 above)
Comment (melekes) [inline cmd/tendermint/commands/root.go]: the sooner we validate the input, the better

File: cmd/tendermint/commands/run_node.go (hunk 8 above)
Comment (melekes) [inline cmd/tendermint/commands/run_node.go]: I will impl it

**Comment by melekes** [comment]
> Yes. Is there ever a reason to not change things to data.Bytes ? I would think anything that might get logged or turned to json should use data.Bytes

```
message ResponseInfo {
	string data = 1;
	string version = 2;
	uint64 last_block_height = 3;
	bytes last_block_app_hash = 4;
}
```

can I use custom types in grpc?

File: cmd/tendermint/commands/root.go (hunk 9 above)
Comment (melekes) [inline cmd/tendermint/commands/root.go]: refactored NewFilterByLevel to return an error

**Pull request merged**